samza-core/src/main/java/org/apache/samza/startpoint/StartpointManager.java

```diff
@@ -173,71 +226,151 @@ public void deleteStartpoint(SystemStreamPartition ssp, TaskName taskName) {
-            writeStartpoint(ssp, taskName, startpoint);
-            sspsToDelete.add(ssp); // Mark for deletion
-            LOG.info("Startpoint for SSP: {} remapped with task: {}.", ssp, taskName);
-          } else {
-            LOG.info("Startpoint for SSP: {} and task: {} already exists and will not be overwritten.", ssp, taskName);
-          }
+    Preconditions.checkArgument(MapUtils.isNotEmpty(taskToSSPs), "taskToSSPs cannot be null or empty");
+
+    // construct fan out with the existing readWriteStore entries and mark the entries for deletion after fan out
+    Instant now = Instant.now();
+    HashMultimap<SystemStreamPartition, TaskName> deleteKeys = HashMultimap.create();
+    HashMap<TaskName, StartpointFanOut> fanOuts = new HashMap<>();
+    for (TaskName taskName : taskToSSPs.keySet()) {
+      Set<SystemStreamPartition> ssps = taskToSSPs.get(taskName);
+      if (CollectionUtils.isEmpty(ssps)) {
+        LOG.warn("No SSPs are mapped to taskName: {}", taskName.getTaskName());
+        continue;
+      }
+      for (SystemStreamPartition ssp : ssps) {
+        Startpoint startpoint = readStartpoint(ssp); // Read SSP-only key
+        if (startpoint != null) {
+          deleteKeys.put(ssp, null);
+        }
+
+        Startpoint startpointForTask = readStartpoint(ssp, taskName); // Read SSP+taskName key
+        if (startpointForTask != null) {
+          deleteKeys.put(ssp, taskName);
+        }
+
+        Startpoint startpointWithPrecedence = resolveStartpointPrecendence(startpoint, startpointForTask);
+        if (startpointWithPrecedence == null) {
+          continue;
+        }
+
+        fanOuts.putIfAbsent(taskName, new StartpointFanOut(now, new HashMap<>()));
+        fanOuts.get(taskName).getFanOuts().put(ssp, startpointWithPrecedence);
+      }
+    }
+
+    if (fanOuts.isEmpty()) {
+      LOG.debug("No fan outs created.");
+      return ImmutableSet.of();
+    }
+
+    LOG.info("Fanning out to {} tasks", fanOuts.size());
+
+    StartpointFanOut.StartpointFanOutSerde startpointFanOutSerde = new StartpointFanOut.StartpointFanOutSerde();
+    for (TaskName taskName : fanOuts.keySet()) {
+      String fanOutKey = toFanOutStoreKey(taskName);
+      StartpointFanOut newFanOut = fanOuts.get(taskName);
+      byte[] fanOutFromStoreInBytes = fanOutStore.get(fanOutKey);
+
+      if (ArrayUtils.isNotEmpty(fanOutFromStoreInBytes)) {
+        // Merge new fan out with existing fan out. This is not a typical scenario but it can happen when the job
+        // restarts before all startpoints are committed by the first checkpoint commit and new startpoints were written
+        // to the readWriteStore after the previous fan out.
+        StartpointFanOut existingFanOut = startpointFanOutSerde.fromBytes(fanOutFromStoreInBytes);
+        existingFanOut.getFanOuts().putAll(newFanOut.getFanOuts());
+        newFanOut = existingFanOut;
+      }
+
+      fanOutStore.put(fanOutKey, startpointFanOutSerde.toBytes(newFanOut));
```

Comment:
The fanout is not transactional across all tasks, so what if there is a failure in the middle of writing all of the fanouts? Could that result in a corrupt state?

Reply:
no exceptions are swallowed, so a failure should cause the caller (i.e. job coordinator) to restart. In that case all the fanouts will be recreated.

Reply:
What if there is a continuous failure after a partial write, so the job coord is not able to finish writing the fanouts? The fanout store would be left in a partial state until the issue is fixed. Would this cause any issue? When the job comes back up, is there a case where the job coord would want to write different fanouts, so that the existing partial ones are no long valid?

Reply:
The partial state would be complete at the subsequent startup of the job coordinator. However, thinking about it more, we probably don't need to do the merge with the existing fan out because the readWriteStore entries would still exist and the fan out's complete state can be rewritten again on the subsequent restart. I'll remove the merge logic.